manual control selection and cleanup

julianoes opened the pull request:
This adds a new module called manual_control which handles as much as possible of RC and MAVLink manual_control input outside of commander.

julianoes commented:
Ok, now it's in a somewhat reviewable state, I hope.

MaEtUgR commented on src/modules/commander/Commander.hpp:
```diff
@@ -252,7 +259,7 @@ class Commander : public ModuleBase<Commander>, public ModuleParams
 		(ParamInt<px4::params::CBRK_VELPOSERR>) _param_cbrk_velposerr,
 		(ParamInt<px4::params::CBRK_VTOLARMING>) _param_cbrk_vtolarming,
 
-		// Geofrence
```
🤣

julianoes commented:
@MaEtUgR @dagar I think now is a good time to review this. Then we can try to merge it once the release is in.

julianoes commented:
Bump, please review.

MaEtUgR reviewed (commented):
I'm on it but it's a bigger junk to review 🚚

MaEtUgR reviewed (commented):
Some more comments, I'm not through yet 🐌

julianoes commented on boards/px4/fmu-v2/default.cmake:
```diff
@@ -70,15 +70,16 @@ px4_add_board(
 		flight_mode_manager
-		fw_att_control
-		fw_pos_control_l1
+		#fw_att_control
```
I'd say so.

julianoes commented on boards/px4/fmu-v2/init/rc.board_mavlink:
```diff
@@ -4,4 +4,4 @@
 
 # Start MAVLink on the USB port
-mavlink start -d /dev/ttyACM0
+mavlink start -d /dev/ttyACM0 -z
```
Oh, this is actually from https://github.com/PX4/PX4-Autopilot/pull/17481 and I no longer depend on it, so I will get rid of it in the next rebase.

Good catch!

julianoes commented on msg/manual_control_input.msg:
```diff
@@ -0,0 +1,52 @@
+uint64 timestamp			# time since system start (microseconds)
```
Gosh I hate how we mix tabs and spaces, and how we align stuff.

MaEtUgR replied:
I agree. We once setteled on tabs for indentation spaces for alignment no? I just realized now that it's mixed in this file but only noticed because things were more misaligned than in `manual_control_setpoint`. Now it looks as good as it gets with the mix.

julianoes commented on msg/manual_control_input.msg:
```diff
@@ -0,0 +1,52 @@
+# The range for the z variable is defined from 0 to 1. (The z field of
+# the MANUAL_CONTROL mavlink message is defined from -1000 to 1000)
+
+float32 x			 # stick position in x direction -1..1
```
Ok, let me try that.

MaEtUgR replied:
Much better 👍

MaEtUgR reviewed (commented):
I studied some more. Still not completely through sorry 🏃

julianoes commented on src/modules/commander/Commander.cpp:
```diff
@@ -841,18 +844,33 @@ Commander::handle_command(const vehicle_command_s &cmd)
 				// Arm is forced (checks skipped) when param2 is set to a magic number.
 				const bool forced = (static_cast<int>(roundf(cmd.param2)) == 21196);
 
+				const bool cmd_from_manual_stick = (static_cast<int>(roundf(cmd.param3)) == 1);
+				const bool cmd_from_manual_switch = (static_cast<int>(roundf(cmd.param3)) == 2);
+				const bool cmd_from_manual_button = (static_cast<int>(roundf(cmd.param3)) == 3);
+				const bool cmd_from_io = (static_cast<int>(roundf(cmd.param3)) == 1234);
```
I guess they should. But also this has already been there before :).

MaEtUgR replied:
I saw that and just like a common definition of such interfaces such that when you do a full-text search you find where it's set/used.

julianoes commented on src/modules/commander/Commander.cpp:
```diff
@@ -868,32 +886,63 @@ Commander::handle_command(const vehicle_command_s &cmd)
 					} else {
-						arming_res = arm(arm_disarm_reason_t::COMMAND_INTERNAL, !forced);
+						if (cmd_from_manual_stick) {
+							arming_res = arm(arm_disarm_reason_t::RC_STICK, !forced);
```
Understood, that's a good idea.

julianoes commented on src/modules/commander/Commander.cpp:
```diff
@@ -868,32 +886,63 @@ Commander::handle_command(const vehicle_command_s &cmd)
 					} else {
-						arming_res = arm(arm_disarm_reason_t::COMMAND_INTERNAL, !forced);
+						if (cmd_from_manual_stick) {
+							arming_res = arm(arm_disarm_reason_t::RC_STICK, !forced);
```
You're right. I'll change that.

julianoes commented on src/modules/commander/Commander.cpp:
```diff
@@ -868,32 +886,63 @@ Commander::handle_command(const vehicle_command_s &cmd)
+							arming_res = arm(arm_disarm_reason_t::RC_STICK, !forced);
+
+						} else if (cmd_from_manual_switch) {
+							arming_res = arm(arm_disarm_reason_t::RC_SWITCH, !forced);
```
Ok.

julianoes commented on src/modules/commander/Commander.cpp:
```diff
@@ -868,32 +886,63 @@ Commander::handle_command(const vehicle_command_s &cmd)
+							arming_res = arm(arm_disarm_reason_t::RC_SWITCH, !forced);
+
+						} else {
+							arming_res = arm(arm_disarm_reason_t::COMMAND_INTERNAL, !forced);
```
I think this one is allowed. For instance if you do `commander arm -f` this is required.

MaEtUgR replied:
Ah yeah, my fault. On the general one it's necessary only by RC it doesn't make sense.

julianoes commented on src/modules/commander/Commander.cpp:
```diff
@@ -2233,146 +2272,70 @@ Commander::run()
-				}
-
-				// evaluate the main state machine according to mode switches
-				if (set_main_state(_status_changed) == TRANSITION_CHANGED) {
```
So set_main_state should be called when a command arrives. Therefore this here is no longer required.

Then, there is IO override which I don't know if it still worked before this PR or was already broken. I actually forgot about that. And I was under the impression that we wanted to get rid of it anyway as part of the IO lobotomy.

MaEtUgR replied:
> So set_main_state should be called when a command arrives.

I guess you mean the main state gets changed when a command gets in right? `set_main_state()` and hence `set_main_state_override_on()` are completely out of the code path with this pr and hence `safety.override_available` and `safety.override_enabled` are also unused. We might as well delete them until https://github.com/PX4/PX4-Autopilot/pull/16444 is in and we implement and test an IO fallback/override that keeps on working.

julianoes replied:
Oh now I get it. I can remove all that stuff.

julianoes commented on src/modules/manual_control/ManualControl.cpp:
```diff
@@ -0,0 +1,575 @@
+		// user wants override
+		const float minimum_stick_change = 0.01f * _param_com_rc_stick_ov.get();
+
+		const bool rpy_moved = (fabsf(_x_diff.diff()) > minimum_stick_change)
```
Good point. I've added a reset method and I'm leaving the diff at 0 for the first iteration, so that should prevent the initial step.

MaEtUgR replied:
Nice!

julianoes commented on src/modules/manual_control/ManualControl.cpp:
```diff
@@ -0,0 +1,575 @@
+
+		_selector.setpoint().user_override = rpy_moved || throttle_moved;
+
+		_x_diff.update(_selector.setpoint().x, dt_s);
```
That's indeed weird. Let me change it.

julianoes commented on src/modules/manual_control/ManualControl.hpp:
```diff
@@ -0,0 +1,180 @@
+public:
+	void update(float value, float dt_s)
+	{
+		if (dt_s < 0.0f) {
+			dt_s = 0.0f;
+
+		} else if (dt_s > _time_period_s) {
+			dt_s = _time_period_s;
+		}
```
Beautiful, thanks :).

julianoes commented on src/modules/manual_control/ManualControl.hpp:
```diff
@@ -0,0 +1,180 @@
+		}
+
+		const float new_diff = value - _last_value;
+		_diff = new_diff * dt_s + _diff * (_time_period_s - dt_s);
```
Uh oh, let me handle that.

julianoes commented on src/modules/manual_control/ManualControl.hpp:
```diff
@@ -0,0 +1,180 @@
+private:
+	float _diff{0.0f};
+	float _last_value{0.0f};
+	const float _time_period_s{1.0f};
```
Right, thanks.

julianoes commented on src/modules/manual_control/ManualControl.hpp:
```diff
@@ -0,0 +1,180 @@
+private:
+	static constexpr int MAX_MANUAL_INPUT_COUNT = 3;
+
+	enum class ArmingAction {
+		TOGGLE = -1,
+		DISARM = 0,
+		ARM = 1,
+	};
+
+	enum class ArmingOrigin {
+		GESTURE = 1,
+		SWITCH = 2,
+		BUTTON = 3,
+	};
```
I agree but I also hate the fact that we don't have type safe enums in uORB and so this is somewhat nicer/safer for this module alone.

MaEtUgR replied:
I'd agree if it was for this module alone but it's the uORB interface which is not typesafe and is used here. The other side in commander has to reinterpret the values with its own hardcoded definition: https://github.com/PX4/PX4-Autopilot/blob/pr-manual_control_selector/src/modules/commander/Commander.cpp#L854-L856

If there was a definition in the message for the arming reasons you would have it common and could e.g. change numbers without adjusting both sides, full-text search for `ARMING_REASON_RC_GESTURE` and immediately see where it's set and distinguished. Also you could just pass it along to the message generation https://github.com/PX4/PX4-Autopilot/blob/pr-manual_control_selector/src/modules/commander/Commander.cpp#L904-L939 instead of handling each case separately.

julianoes replied:
Alright, done.

julianoes commented on src/modules/commander/Commander.cpp:
```diff
@@ -2233,146 +2272,70 @@ Commander::run()
-			    && _manual_control.wantsOverride(_vehicle_control_mode)) {
+			    && _armed.armed
+			    && !_status_flags.rc_input_blocked
+			    && manual_control_setpoint.user_override
```
Oh, nice catch. I think it makes sense to check just in case.

julianoes commented:
@MaEtUgR thanks a lot for the in-depth review.

I think I have addressed most points, except:
- Adding enums to the uORB message. I don't love the uORB enums because they are not type safe, so it's a trade-off.
- Checking the IO override stuff. I believe this is either broken already or will be broken with the IO lobotomy. :confused: 

I've rebased the branch and force pushed, so to review the changes, best just check the recent commits.

MaEtUgR reviewed (commented):
Just adding this now already in the hope it's what prohibits me from answering in the existing conversations.

dagar commented on src/modules/manual_control/ManualControlSelector.cpp:
```diff
@@ -0,0 +1,121 @@
+	setpoint.chosen_input.y = input.y;
+	setpoint.chosen_input.z = input.z;
+	setpoint.chosen_input.r = input.r;
+	// FIXME: what's that?
+	//setpoint.chosen_input.vx = (input.x - _manual_control_input[i].x) * dt_inv;
+	//setpoint.chosen_input.vy = (input.y - _manual_control_input[i].y) * dt_inv;
+	//setpoint.chosen_input.vz = (input.z - _manual_control_input[i].z) * dt_inv;
+	//setpoint.chosen_input.vr = (input.r - _manual_control_input[i].r) * dt_inv;
```
This is just something on the side I was playing with initially for things like deciding if the sticks are moving. Now that all input sources are routed through this module we can track the update rates. This could be deleted for now.

MaEtUgR commented on src/modules/manual_control/ManualControlSelectorTest.cpp:
```diff
@@ -0,0 +1,175 @@
+using namespace time_literals;
+using namespace manual_control;
+
+static constexpr uint64_t some_time = 1234568;
```
```suggestion
static constexpr uint64_t some_time = 12345678;
```
😛

julianoes replied:
No way :facepalm:

MaEtUgR commented on src/modules/manual_control/ManualControlSelectorTest.cpp:
```diff
@@ -0,0 +1,198 @@
+	EXPECT_EQ(selector.instance(), -1);
+}
+
+TEST(ManualControlSelector, RcOutdated)
```
I added this test to show that you could end up accepting outdated data if it arrives already being older than the timeout. That's happening because checking the timeout only happens for the signal that's already in use so you end up invalidating the old outdated one and accepting the new outdated one in every iteration.

julianoes replied:
Good catch. Fixed it and test is now passing :tada:.

MaEtUgR commented on src/modules/mavlink/mavlink_receiver.cpp:
```diff
@@ -1910,6 +1868,7 @@ MavlinkReceiver::handle_message_rc_channels_override(mavlink_message_t *msg)
 	}
 
 	// publish uORB message
+	rc.timestamp = hrt_absolute_time();
```
Only nees one hrt call since it's anyways the same.
```suggestion
	rc.timestamp = rc.timestamp_last_signal = hrt_absolute_time();
```

MaEtUgR commented on src/modules/mavlink/mavlink_receiver.cpp:
```diff
@@ -1861,16 +1826,9 @@ MavlinkReceiver::handle_message_rc_channels_override(mavlink_message_t *msg)
-	// metadata
-	rc.timestamp = hrt_absolute_time();
-	rc.timestamp_last_signal = rc.timestamp;
+	rc.timestamp_last_signal = hrt_absolute_time();
```
```suggestion
```

MaEtUgR commented on src/modules/mavlink/mavlink_receiver.cpp:
```diff
@@ -1924,55 +1883,15 @@ MavlinkReceiver::handle_message_manual_control(mavlink_message_t *msg)
+	manual.r = man.r / 1000.0f;
+	manual.z = man.z / 1000.0f;
+	manual.data_source = manual_control_input_s::SOURCE_MAVLINK_0 + _mavlink->get_instance_id();
+	manual.timestamp = hrt_absolute_time();
```
```suggestion
	manual.timestamp = manual.timestamp_sample = hrt_absolute_time();
```

MaEtUgR commented on src/modules/mavlink/mavlink_receiver.cpp:
```diff
@@ -1924,55 +1883,15 @@ MavlinkReceiver::handle_message_manual_control(mavlink_message_t *msg)
-		_manual_control_setpoint_pub.publish(manual);
-	}
+	manual_control_input_s manual{};
+	manual.timestamp_sample = hrt_absolute_time();
```
```suggestion
```

MaEtUgR commented on src/modules/manual_control/ManualControl.cpp:
```diff
@@ -0,0 +1,578 @@
+
+					if (switches.gear_switch != _previous_switches.gear_switch) {
+
+						if (switches.gear_switch == manual_control_switches_s::SWITCH_POS_ON) {
+							publish_landing_gear(landing_gear_s::GEAR_UP);
+
+						} else if (switches.gear_switch == manual_control_switches_s::SWITCH_POS_OFF) {
+							publish_landing_gear(landing_gear_s::GEAR_DOWN);
+						}
```
Maybe we could solve it by just looking at the `manual_control_switches.switch_changes`. The value needs to increase after (re)gaining RC for something to happen.

julianoes replied:
> don't allow overriding gear when in autonomous mode -> now possible, changed question

I think that's how it was before, see:
https://github.com/PX4/PX4-Autopilot/blob/dddd2c3297bd57a92d3b0261e2702f4f36d8a774/src/modules/commander/Commander.cpp#L2309

> don't change gear state when the switch is already in a certain position on initialization -> I think this got lost, it's why there was last_gear_switch != NONE in the condition before x

Good catch! I think I can just add it back in, like it was before.

MaEtUgR reviewed (commented):
Now I'm completely through 😅 I hope the feedback helps. I'm happy to cooperate e.g. on the arm button logic. I added two commits for the unit test I came up with that fails and to remove the commented-out lines.
My main concerns are:
- arming button "broken"
- gear switch doesn't cover all cases
- MAVLink command target IDs

julianoes commented on src/modules/manual_control/ManualControl.cpp:
```diff
@@ -0,0 +1,578 @@
+			_previous_arm_gesture = true;
+			send_arm_command(ArmingAction::ARM, ArmingOrigin::GESTURE);
+
+		} else if (!_selector.setpoint().arm_gesture) {
+			_previous_arm_gesture = false;
+		}
```
Right, fixed as you suggested. I don't think it makes a difference though.

julianoes commented on src/modules/manual_control/ManualControl.cpp:
```diff
@@ -0,0 +1,578 @@
+				       || (fabsf(_r_diff.diff()) > minimum_stick_change);
+
+		// Throttle change value doubled to achieve the same scaling even though the range is [0,1] instead of [-1,1]
+		const bool throttle_moved = (fabsf(_z_diff.diff()) * 2.f > minimum_stick_change);
```
Yes, ok.

julianoes commented on src/modules/manual_control/ManualControl.cpp:
```diff
@@ -0,0 +1,578 @@
+						if (_param_com_arm_swisbtn.get()) {
+							// Arming button
+							if (switches.arm_switch == manual_control_switches_s::SWITCH_POS_ON) {
+								send_arm_command(ArmingAction::TOGGLE, ArmingOrigin::BUTTON);
```
Hm, good point. I should add that back in.

julianoes commented on src/modules/mc_rate_control/MulticopterRateControl.cpp:
```diff
@@ -153,8 +153,14 @@ MulticopterRateControl::Run()
 			if (_landing_gear_sub.copy(&landing_gear)) {
-				if (landing_gear.landing_gear != landing_gear_s::GEAR_KEEP) {
-					_landing_gear = landing_gear.landing_gear;
+				if (landing_gear.landing_gear != landing_gear_s::GEAR_KEEP &&
+				    _v_control_mode.flag_control_manual_enabled) {
```
I think I agree but here I was just trying to preserve how it was:
https://github.com/PX4/PX4-Autopilot/blob/dddd2c3297bd57a92d3b0261e2702f4f36d8a774/src/modules/commander/Commander.cpp#L2309

MaEtUgR replied:
ok, but it's different now. Trying to simplify it was:
```
Commander: if (manual_mode && landed) { publish gear from switch }
RateControl: actuation = subscribed gear
```
and now it is:
```
ManualControl: publish gear switch
RateControl: if (manual_mode && landed) { actuation = subscribed gear }
```

As a result if I understood correctly:
- any autonomous mode cannot move the gear anymore because its publication gets ignored
- in autonomous modes you can now move the gear by switch (even when landed)

julianoes replied:
I suggest to enable the gear always to be changed unless we're landed and want to put it up.

julianoes commented on src/modules/manual_control/ManualControl.cpp:
```diff
@@ -0,0 +1,578 @@
+					}
+
+				} else {
+					// Send an initial command to switch to the mode requested by R
```
:blush:

julianoes commented on src/modules/manual_control/ManualControl.cpp:
```diff
@@ -0,0 +1,578 @@
+				_previous_switches = switches;
+
+			} else {
+				_previous_switches = {};
```
Right!

julianoes commented on src/modules/manual_control/ManualControl.cpp:
```diff
@@ -0,0 +1,578 @@
+
+				} else {
+					// Send an initial command to switch to the mode requested by R
+					evaluate_mode_slot(switches.mode_slot);
```
Arg, so many corner cases.

julianoes commented on src/modules/manual_control/ManualControl.hpp:
```diff
@@ -0,0 +1,191 @@
+
+	ManualControlSelector _selector;
+	bool _published_invalid_once{false};
+	int _last_selected_input{-1};
```
Nice!

julianoes commented on src/modules/manual_control/ManualControl.cpp:
```diff
@@ -0,0 +1,578 @@
+		return;
+	}
+
+	command.target_system = 1;
+	command.target_component = 1;
```
Wow that was sloppy, I will get it from the param.

julianoes commented on src/modules/manual_control/ManualControlSelector.hpp:
```diff
@@ -0,0 +1,62 @@
+public:
+	void set_rc_in_mode(int32_t rc_in_mode) { _rc_in_mode = rc_in_mode; }
+	void set_timeout(uint64_t timeout) { _timeout = timeout; }
+	void update_time_only(uint64_t now);
```
I find that more confusing because I would assume `update_timeout` to be basically the same as `set_timeout`.

julianoes commented on src/modules/manual_control/ManualControl.cpp:
```diff
@@ -0,0 +1,578 @@
+			_last_selected_input = _selector.instance();
+		}
+
+		_manual_control_input_subs[_selector.instance()].registerCallback();
```
Good point. I'm gonna add a check just in case.

julianoes commented on src/modules/manual_control/ManualControlSelector.cpp:
```diff
@@ -0,0 +1,121 @@
+		}
+
+	} else {
+		// FIXME: param value unknown, what to do?
```
Good idea. Except I'm going to print an error and mark it as invalid.

julianoes commented on src/modules/manual_control/ManualControlSelector.cpp:
```diff
@@ -0,0 +1,121 @@
+		}
+
+	} else if (_rc_in_mode == 2) {
+		// FIXME: what to do in the legacy case?
```
Removed it.

julianoes commented on src/modules/manual_control/ManualControlSelector.cpp:
```diff
@@ -0,0 +1,121 @@
+
+	if (_rc_in_mode == 0 && input.data_source == manual_control_input_s::SOURCE_RC) {
+		_setpoint = setpoint_from_input(input);
+		_setpoint.valid = true;
+		_instance = instance;
```
Not 100% convinced but good idea.

julianoes commented:
@MaEtUgR I think I've answered to most of your points and fixed most of them. I'm also rebasing this on latest master, so to see if my fixes make sense, you can just check the latest commits.

cmic0 commented:
Something worth to incorporate as an incremental PR on top of this one is the toggle buttons over S.BUS feature, https://github.com/PX4/PX4-Autopilot/pull/17270. Sadly I let it go rotten but hopefully we can get it in in a better way with this re-structuring :)

bkueng reviewed (commented):
Nice improvement in general

julianoes commented on src/modules/manual_control/ManualControl.cpp:
```diff
@@ -0,0 +1,588 @@
+	command_pub.publish(command);
+}
+
+void ManualControl::send_arm_command(int8_t action, int8_t origin)
```
Valid concerns. @dagar what do you think?

julianoes commented on src/modules/manual_control/ManualControl.cpp:
```diff
@@ -0,0 +1,588 @@
+		const int instance = _selector.instance();
+
+		if (instance >= 0 && instance < MAX_MANUAL_INPUT_COUNT) {
+			_manual_control_input_subs[instance].registerCallback();
```
I'm not sure how that is supposed to work :confused:.

julianoes commented on msg/manual_control_setpoint.msg:
```diff
@@ -1,49 +1,10 @@
+uint64 timestamp                        # time since system start (microseconds)
 
-uint64 timestamp_sample                # the timestamp of the raw data (microseconds)
+px4/manual_control_input chosen_input
```
FYI @MaEtUgR

MaEtUgR commented on src/modules/commander/state_machine_helper.cpp:
```diff
@@ -368,9 +368,13 @@ main_state_transition(const vehicle_status_s &status, const main_state_t new_mai
 
 	case commander_state_s::MAIN_STATE_AUTO_MISSION:
 
+		if (!status_flags.condition_auto_mission_available) {
```
What does this change? From above you can see `ret` stays `TRANSITION_DENIED` if it's not changed which means it was already denied if you had no valid mission.

MaEtUgR commented on src/modules/manual_control/ManualControlSelector.cpp:
```diff
@@ -0,0 +1,121 @@
+	// This method requires the current timestamp explicitly in order to prevent the case
+	// where the timestamp_sample of some source is already outdated.
+
+	if (now >= input.timestamp_sample && now - input.timestamp_sample > _timeout) {
+		_setpoint.valid = false;
+		_instance = -1;
+		return;
+	}
```
Doesn't this mean that any invalid input would immediately invalidate another chosen and valid input?

MaEtUgR commented:
I'm rebasing again on https://github.com/PX4/PX4-Autopilot/pull/18168

MaEtUgR commented:
I rebased on master and rewrote the history to make sure the individual commits make sense and build. There were some nasty conflict artifacts in there.

dagar commented:
> I rebased on master and rewrote the history to make sure the individual commits make sense and build. There were some nasty conflict artifacts in there.

Feel free to squash my original commit out of existence, I'd rather not see it end up in master.

![Screenshot from 2021-11-01 12-21-37](https://user-images.githubusercontent.com/84712/139705205-4dde1758-cc2b-43e5-934e-127b8f981f8e.png)

MaEtUgR commented:
During real-world testing I sadly found some more issues. I'm working through them.
Mode initialization and report for modes which are not available I already fixed but I still saw some problems with RC override not being properly time abstracted and the selection. Continuing my work. Since I'm doing changes again I thought about making `manual_control_setpoint` backwards compatible because I don't see much added value in having a separate message containing basically just the other one as `chosen_input`.

julianoes commented:
> Since I'm doing changes again I thought about making manual_control_setpoint backwards compatible because I don't see much added value in having a separate message containing basically just the other one as chosen_input.

I don't understand. Without that additional message you don't actually have the separation between what is coming in what is selected, right?

MaEtUgR commented:
> I don't understand. Without that additional message you don't actually have the separation between what is coming in what is selected, right?

You have the separation in topics but not in message because the message is almost the same. I committed my suggestion.

julianoes commented:
I preferred it as it was before because it makes it harder to misuse but I'm fine with it.

MaEtUgR commented:
Final (?) rebase on `px4/master` because of a simple conflict with https://github.com/PX4/PX4-Autopilot/pull/17547/files#diff-32ce7fd2aab8a5ba5092fe880ac2055b0a153914e38b40aad6cadffb811fd7d0R73

MaEtUgR reviewed (approved):
Jenkins fails on `src/drivers/uavcan/libuavcan` and these targets build for me locally.
I'm done with my testing.
@julianoes @dagar Merge?
